File: packages/core/src/types/TableStore.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type ColumnTable from 'arquero/dist/types/table/column-table'

/**
 * Resolver function that looks up a table by id.
 */
export type ResolverFunction = (id: string) => Promise<ColumnTable>

/**
 * Function callback for table change listeners.
 */
export type ListenerFunction = (containre: TableContainer) => void

/**
 * Function callback for general activity listener.
 */
export type ChangeListenerFunction = () => void

export interface TableContainer {
	/**
	 * This is the formal id for a table, and must be unique within the store.
	 * A URI would normally be appropriate.
	 */
	id: string
	/**
	 * This is an optional alias or friendly name for the table.
	 */
	name?: string
	/**
	 * This is the actual Arquero table instance to store.
	 * If it has not been resolved yet it will be undefined.
	 */
	table?: ColumnTable
	/**
	 * Optional resolver function to lazy-load a table when first requested.
	 * If a table is not found, the resolver will be invoked or an error thrown.
	 */
	resolver?: ResolverFunction
}

/**
 * Store for a collection of tables, used as an execution storage context for pipelines.
 */
export interface TableStore {
	/**
	 * Returns a table container.
	 * Uses async resolver function if necessary to lazy-load or retrieve remote tables.
	 * @param id
	 */
	get(id: string): Promise<TableContainer>
	table(id: string): Promise<ColumnTable>
	/**
	 * Set a loaded table in the store.
	 * @param id
	 * @param table
	 */
	set(container: TableContainer): void
	/**
	 * Remove the named table
	 * @param id
	 */
	delete(id: string): void
	/**
	 * Add a table name to the store with a resolver function to be loaded when needed.
	 * @param id
	 * @param resolver
	 */
	queue(id: string, resolver: ResolverFunction): void
	/**
	 * List all tables in the store by id, with an optional filter function.
	 * @param filter
	 * @returns
	 */
	list(filter?: (id: string) => boolean): string[]
	/**
	 * Resolves all tables and converts to a id:table Map
	 */
	toMap(): Promise<Map<string, TableContainer>>
	/**
	 * Resolves all tables and convers to an array
	 */
	toArray(): Promise<TableContainer[]>
	/**
	 * Add a listener for a particular table.
	 * @param id
	 * @param listener
	 */
	listen(id: string, listener: ListenerFunction): void
	/**
	 * Stop listening for a particular table.
	 * @param id
	 */
	unlisten(id: string): void
	/**
	 * Get alerted for any changes in the store.
	 * @param listener
	 */
	addChangeListener(listener: ChangeListenerFunction): void
	/**
	 * Print the whole store to the console.
	 */
	print(): Promise<void>
	/**
	 * Deeps clones the tables in this store for creating sub-contexts, etc.
	 * @returns
	 */
	clone(): Promise<TableStore>
	/**
	 * Deletes all tables from store
	 * @returns
	 */
	clear(): void
}

File: python/data_wrangling_components/table_store.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

from typing import Callable, Dict, List, Optional, Union

import pandas as pd

from dataclasses import dataclass, field
from pandas.core.groupby import DataFrameGroupBy

from data_wrangling_components.exceptions import (
    ResolverNotFoundError,
    TableNotFoundError,
)


@dataclass
class Table:
    name: str
    data: Optional[Union[pd.DataFrame, DataFrameGroupBy]] = None
    resolver: Optional[Callable[[str], pd.DataFrame]] = None

    def deepcopy(self):
        return Table(self.name, self.data.copy(), self.resolver)


@dataclass
class TableStore:
    _tables: Dict[str, Table] = field(default_factory=dict)

    def get(self, name: str) -> Union[pd.DataFrame, DataFrameGroupBy]:
        """Retrieves a dataframe from the table store

        :param name:
            Name of the table to retrieve
        :type name: str
        :raises TableNotFoundError:
            Raised when name does not exist in table
        :raises ResolverNotFoundError:
            Raised when the resolver is not found for that table
        :return: the dataframe with the data
        :rtype: Union[pd.DataFrame, DataFrameGroupBy]
        """
        try:
            container = self._tables[name]
        except KeyError:
            raise TableNotFoundError(f"No table named '{name}' found in store.")

        table = container.data

        if table is None:
            resolver_func = container.resolver
            if resolver_func is None:
                raise ResolverNotFoundError(
                    f"No resolver function for unloaded table '{name}'."
                )
            table = container.resolver(name)
            self._tables[name] = table

        return table

    def set(self, name: str, table: pd.DataFrame) -> None:
        """Sets the table for a name in the store

        :param name:
            The name of the table.
            If the name already exists then the data is overwritten
        :type name: str
        :param table:
            The DataFrame with the data
        :type table: pd.DataFrame
        """
        self._tables[name] = Table(name=name, data=table)

    def delete(self, name: str) -> None:
        """Deletes a table from the store

        :param name: Name of the table to delete
        :type name: str
        """
        self._tables.pop(name)

    def queue(self, name: str, resolver: Callable[[str], pd.DataFrame]) -> None:
        """Queues a table to be lazy loaded with a resolver function

        :param name:
            Name of the table to be lazy loaded
        :type name: str
        :param resolver:
            The function to be used to load the table
        :type resolver: Callable[[str], pd.DataFrame]
        """
        self._tables[name] = Table(name=name, resolver=resolver)

    def list(self, filter: Optional[str] = None) -> List[str]:
        """Lists all tables stored in the store

        :param filter:
            Optional filter to reduce the amount of tables listed, defaults to None
        :type filter: Optional[str], optional
        :return:
            The list of table names stored in the store
        :rtype: List[str]
        """
        return [
            key for key in list(self._tables.keys()) if filter is None or filter in key
        ]

    def to_map(self) -> Dict[str, pd.DataFrame]:
        """Creates a map of name: DataFrame from the store

        :return:
            A Dictionary with name as key and DataFrame as value
        :rtype: Dict[str, pd.DataFrame]
        """
        return {table.name: self.get(table.name) for table in self._tables.values()}

    def deepcopy(self):
        return TableStore(
            {name: table.deepcopy() for name, table in self._tables.items()}
        )
